src/wp-includes/block-template-utils.php

```diff
@@ -88,26 +596,96 @@ function get_block_templates( $query = array(), $template_type = 'wp_template' )
 
-		if ( ! is_wp_error( $template ) ) {
-			$query_result[] = $template;
+		if ( is_wp_error( $template ) ) {
+			continue;
+		}
+
+		if ( $post_type && ! $template->is_custom ) {
+			continue;
 		}
+
+		$query_result[] = $template;
 	}
 
-	return $query_result;
+	if ( ! isset( $query['wp_id'] ) ) {
+		$template_files = _get_block_templates_files( $template_type );
+		foreach ( $template_files as $template_file ) {
+			$template = _build_block_template_result_from_file( $template_file, $template_type );
+
+			if ( $post_type && ! $template->is_custom ) {
+				continue;
+			}
+
+			if ( $post_type &&
+				isset( $template->post_types ) &&
+				! in_array( $post_type, $template->post_types, true )
+			) {
+				continue;
+			}
+
+			$is_not_custom   = false === array_search(
+				wp_get_theme()->get_stylesheet() . '//' . $template_file['slug'],
+				array_column( $query_result, 'id' ),
+				true
+			);
+			$fits_slug_query =
+				! isset( $query['slug__in'] ) || in_array( $template_file['slug'], $query['slug__in'], true );
+			$fits_area_query =
+				! isset( $query['area'] ) || $template_file['area'] === $query['area'];
+			$should_include  = $is_not_custom && $fits_slug_query && $fits_area_query;
+			if ( $should_include ) {
+				$query_result[] = $template;
+			}
+		}
+	}
+
+	/**
+	 * Filters the array of queried block templates array after they've been fetched.
+	 *
+	 * @since 5.9
+	 *
+	 * @param WP_Block_Template[] $query_result Array of found block templates.
+	 * @param array $query {
+	 *     Optional. Arguments to retrieve templates.
+	 *
+	 *     @type array  $slug__in List of slugs to include.
+	 *     @type int    $wp_id Post ID of customized template.
+	 * }
+	 * @param array $template_type wp_template or wp_template_part.
+	 */
+	return apply_filters( 'get_block_templates', $query_result, $query, $template_type );
 }
 
 /**
  * Retrieves a single unified template object using its id.
  *
  * @since 5.8.0
  *
- * @param string $id            Template unique identifier (example: theme_slug//template_slug).
- * @param string $template_type Optional. The template type (post type). Default 'wp_template'.
+ * @param string $id Template unique identifier (example: theme_slug//template_slug).
+ * @param array  $template_type wp_template or wp_template_part.
+ *
  * @return WP_Block_Template|null Template.
  */
 function get_block_template( $id, $template_type = 'wp_template' ) {
+	/**
+	 * Filters the block templates array before the query takes place.
+	 *
+	 * Return a non-null value to bypass the WordPress queries.
+	 *
+	 * @since 5.9
+	 *
+	 * @param WP_Block_Template|null $block_template Return block template object to short-circuit the default query,
+	 *                                               or null to allow WP to run it's normal queries.
+	 * @param string $id Template unique identifier (example: theme_slug//template_slug).
+	 * @param array  $template_type wp_template or wp_template_part.
```

Comment:
```suggestion
	 * @since 5.9.0
	 *
	 * @param WP_Block_Template|null $block_template Return block template object to short-circuit the default query,
	 *                                               or null to allow WP to run its normal queries.
	 * @param string $id                             Template unique identifier (example: theme_slug//template_slug).
	 * @param array  $template_type                  Template type: `'wp_template'` or '`wp_template_part'`.
```